pkg/madmin/kms-commands.go

```diff
@@ -59,3 +61,54 @@ type KMSKeyStatus struct {
 	EncryptionErr string `json:"encryption-error,omitempty"` // An empty error == success
 	DecryptionErr string `json:"decryption-error,omitempty"` // An empty error == success
 }
+
+// RotateKeys performs a key rotation of the data encryption keys (DEK) derived by
+// the KMS. The MinIO server will inspect all objects specified by req and requests
+// a new data encryption key from the KMS for this object.
+//
+// RotateKeys returns a stream of newline-delimited JSON objects. Each object represents
+// one successful key rotation.
+func (adm *AdminClient) RotateKeys(req *KMSRotateKeyRequest) (io.ReadCloser, error) {
```

Comment:
Rather than returning `io.ReadCloser` can we not send a channel of outputs?  for example below code should be ideally in this API

```
	scanner := bufio.NewScanner(resp)
￼	for scanner.Scan() {
￼		var response madmin.KMSRotateKeyResponse
￼		if err = json.Unmarshal(scanner.Bytes(), &response); err != nil {
￼			log.Fatalln(err)
￼		}
￼		fmt.Println(response.JSON())
￼	}
￼```

Reply:
This has been done since the server eventually sends a lot of JSON to the client. When the client wants to output the JSON (e.g. `--json`) unmarshal the JSON stream just to marshal it again may lead to issues when the client cannot keep up with the server. Note that:
```
for scanner.Scan() {
	fmt.Println(scanner.Text())
}
```
would pirnt exactly the same.

Callers that require a struct can unmarshal explicitly.
However, we may even avoid streaming everything to client in general. Have to think about this since that makes the server implementation significantly more complex.

Reply:
It won't make server implementation complex, all you are doing is sending a channel upwards reading like by line. Sending a reader looks weird IMO, this API should be similar to Listing which returns a struct of that channel.

Consumption by `mc` becomes very easy. Also we don't need to write this API just to satisfy mc it should be a proper API so that a user knows what to do with the returned output. A returned channel is more natural than a byte stream.

Reply:
> It won't make server implementation complex, all you are doing is sending a channel upwards reading line by line.

Yeah, it is clear that such a client-side API does not affect the server. That's not what I mean.
I referred to a different server side implementation that does not send a stream on JSON objects (nd-json) but accepts the client request and does the rotation as background process. The client then would have an API to query the status of this background job. (@klauspost suggested something like this). This avoids sending a (potentially) large stream to the client. (Async API).
However, it makes the server implementation much more complex since we have to think about what happens when the server dies / gets shutdown a.s.o.

> Sending a reader looks weird IMO, this API should be similar to Listing which returns a struct of that channel.

This has been done b/c of performance concerns. However, with either simd-json or a different API this is not an issue. Therefore, we can have a channel (or even better a iterator-type like `bufio.Scanner`).

> Consumption by mc becomes very easy. Also we don't need to write this API just to satisfy mc it should be a proper API so that a user knows what to do with the returned output.

That's the goal anyway. I prefer a clean & idiomatic (server/client) API.